docs/Toy/toy/rewrites/inline_toy.py

```diff
@@ -0,0 +1,102 @@
+from xdsl.ir import MLContext
+from xdsl.dialects.builtin import (
+    ModuleOp,
+    StringAttr,
+)
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriteWalker,
+    op_type_rewrite_pattern,
+    RewritePattern,
+    PatternRewriter,
+)
+from xdsl.transforms.dead_code_elimination import dce
+
+from ..dialects import toy
+
+
+class InlineFunctions(RewritePattern):
+    _func_op_by_name: dict[str, toy.FuncOp] | None = None
+
+    def func_op(self, module: ModuleOp, name: str) -> toy.FuncOp:
+        if self._func_op_by_name is None:
+            self._func_op_by_name = {
+                op.sym_name.data: op for op in module.ops if isinstance(op, toy.FuncOp)
+            }
+        return self._func_op_by_name[name]
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: toy.GenericCallOp, rewriter: PatternRewriter):
+        """
+        For each generic call, find the function that it calls, and inline it.
+        """
+
+        # Get module
+        parent = op.parent_op()
+        assert isinstance(parent, toy.FuncOp)
+        module = parent.parent_op()
+        assert isinstance(module, ModuleOp)
+
+        # Clone called function
+        impl = self.func_op(module, op.callee.string_value()).clone()
+        impl_block = impl.body.block
+
+        # Cast operands to unranked
+        inputs = [toy.CastOp(operand) for operand in op.operands]
+
+        # Insert casts before matched op
+        rewriter.insert_op_before_matched_op(inputs)
+
+        # Replace block args with operand casts
+        for i, arg in zip(inputs, impl_block.args):
+            arg.replace_by(i.res)
+
+        # remove block args
+        while len(impl_block.args):
+            assert not len(impl_block.args[-1].uses)
+            rewriter.erase_block_argument(impl_block.args[-1])
+
+        # Inline function definition before matched op
+        rewriter.inline_block_before_matched_op(impl_block)
+
+        # Get return from function definition
+        return_op = op.prev_op
+        assert return_op is not None
+
+        rewriter.replace_matched_op([], return_op.operands)
+        rewriter.erase_op(return_op)
+
+
+class RemoveUnusedPrivateFunctions(RewritePattern):
+    _used_funcs: set[str] | None = None
+
+    def should_remove_op(self, op: toy.FuncOp) -> bool:
+        if op.sym_visibility != StringAttr("private"):
+            return False
```

Comment:
It would be very cool to have these kinds of checks as part of the rewrite pattern checks at some point :)

Reply:
I'm pretty sure Sasha would agree with you ;)
I'm not sure what's your plan on this though, I don't really see how to make this easier :/

Reply:
Yep, will hopefully have a working prototype to discuss soon